candle-core/src/metal_backend.rs

```diff
@@ -126,30 +268,48 @@ impl BackendStorage for MetalStorage {
         let el = shape.elem_count();
         let dtype = self.dtype;
 
-        if layout.is_contiguous() || layout.start_offset() != 0 || dtype != DType::F32 {
-            crate::bail!("Not contiguous, non-f32 affine is not implemented yet.");
+        let buffer = device.new_buffer(el, self.dtype);
+        let command_buffer = self.device.command_buffer();
+        if layout.is_contiguous() && layout.start_offset() == 0 {
+            let name = match self.dtype {
+                DType::F32 => "affine_float",
+                DType::F16 => "affine_half",
+                dtype => crate::bail!("Affine {dtype:?}"),
+            };
+            candle_metal_kernels::call_affine(
+                &device.device,
+                &command_buffer,
+                &device.kernels,
+                name,
+                el,
+                &self.buffer,
+                &buffer,
+                mul as f32,
+                add as f32,
+            )
+            .map_err(MetalError::from)?;
```

Comment:
Feels a bit that these error conversions could be handled by the `?` operator. Is there some difficulty in implementing the appropriate conversion traits so that we can remove these `map_err`?

Reply:
Because our big error type doesn't implement those froms.

We could implement the from `MetalKernelError` directly in our large error sum type, but I think it's better to encapsulate everything in raw `MetalError` which encapsulates everything linked to metal (Kernel and non kernel errors).

I don't like those map_err either, but I don't know of any simpler way to do this hierarchy (without modifyin the device trait).